letslambda.py

```diff
@@ -142,127 +348,217 @@ def answer_dns_challenge(client, domain, challenge):
+        return None
+
+    add_status = wait_letsencrypt_record_insync(conf, add_status)
+    if add_status == None:
+        LOG.error("Cannot determine if the dns record has been correctly created. Skipping domain {0}".format(domain['name']))
+        return None
+
+    if add_status == False:
+        LOG.error("We updated R53 but the servers didn't sync within 60 seconds. Skipping domain {0}".format(domain['name']))
+        return None
+
+    if add_status is not True:
+        LOG.error("An unexpected result code has been returned. Please report this bug. Skipping domain {0}".format(domain['name']))
+        LOG.error("add_status={0}".format(add_status))
+        return None
 
     ## Now, let's tell the ACME server that we are ready
     challenge_response = challenges.DNS01Response(key_authorization=authorization)
     challenge_resource = client.answer_challenge(challenge, challenge_response)
 
-    # We'd better poll to get the answer of this.
+    if challenge_resource.body.error != None:
+        return False
 
-def generateCSR(bucket, conf):
-    if 'reuse_key' in conf:
-        key = loadCSRKey(bucket, conf)
-    else:
-        key = crypto.PKey()
-        key.generate_key(crypto.TYPE_RSA, 4096)
+    return True
 
-    csr = crypto.X509Req()
-    csr.get_subject().countryName = conf['countryName']
-    csr.get_subject().CN = conf['name']
-    csr.set_pubkey(key)
-    csr.sign(key, "sha1")
-    return (csr, key)
+def create_and_save_key(conf, s3_key, kms_key='AES256'):
+    """
+    Generate a RSA 4096 key for general purpose (account or CSR)
+    """
+    LOG.info("Generating new RSA key")
+    key = RSA.generate(4096).exportKey("PEM")
+    save_to_s3(conf, s3_key, key, True, kms_key)
+    return key
 
-def requestCertificate(client, bucket, conf, auth_resource):
-    (csr, key) = generateCSR(bucket, conf)
-    # Gently wait for authorization to be valid
-    # wait_until_valid(auth_resource)
+def save_to_s3(conf, s3_key, content, encrypt=False, kms_key='AES256'):
+    """
+    Save the rsa key in PEM format to s3 .. for later use
+    """
+    LOG.info("Saving object '{0}' to in 's3://{1}'".format(s3_key, conf['s3_bucket']))
+    s3 = conf['s3_client']
+    try:
+        if encrypt == True:
+            if  kms_key != 'AES256':
+                s3.put_object(Bucket=conf['s3_bucket'],
+                        Key=s3_key,
+                        Body=content,
+                        ACL='private',
+                        ServerSideEncryption='aws:kms',
+                        SSEKMSKeyId=kms_key)
+            else:
+                s3.put_object(Bucket=conf['s3_bucket'],
+                        Key=s3_key,
+                        Body=content,
+                        ACL='private',
+                        ServerSideEncryption='AES256')
+        else:
+                s3.put_object(Bucket=conf['s3_bucket'],
+                        Key=s3_key,
+                        Body=content,
+                        ACL='private')
```

Comment:
The above, is a clear example of kwargs creation  ? can we get that instead ? This will simplify the code.

```
kwargs = {'Bucket': ..., 'Key': ...}
if encrypt == True and kms_key == 'AES256':
  kwargs['ServerSideEncryption'] = "AES256'
s3.put_object